File: framework/languages/java/samples/java/SupportChat/Server/Support/src/main/java/systems/zlink/samples/supportchat/server/support/infrastructure/zlink/spots/conversationspot/notifications/ConversationNotificationPublisher.java
package systems.zlink.samples.supportchat.server.support.infrastructure.zlink.spots.conversationspot.notifications;

import java.util.List;
import java.util.Map;
import java.util.function.Consumer;
import systems.zlink.samples.supportchat.server.support.infrastructure.zlink.actors.SupportUserActor;
import systems.zlink.samples.supportchat.server.support.domain.conversation.ConversationModels.ConversationEvent;
import systems.zlink.samples.supportchat.server.support.domain.conversation.ConversationModels.Roles;
import systems.zlink.samples.supportchat.shared.contracts.Messages;

public final class ConversationNotificationPublisher {
    public void publish(
        List<ConversationEvent> events,
        Map<String, SupportUserActor> actors) {
        for (ConversationEvent event : events) {
            publish(event, actors);
        }
    }

    public void publishJoinedAgentToCustomer(
        SupportUserActor customer,
        Messages.ConversationState state) {
        if (state.agentActorId() == null) {
            return;
        }
        customer.context().boundSession()
            .send(new Messages.ParticipantJoinedNotify(
                state.conversationId(),
                state.agentActorId(),
                Roles.Agent,
                state))
            .await();
    }

    private void publish(ConversationEvent event, Map<String, SupportUserActor> actors) {
        switch (event.kind()) {
            case PARTICIPANT_JOINED -> publishParticipantJoined(event, actors);
            case ASSIGNED -> publishAssigned(event, actors);
            case MESSAGE_APPENDED -> publishMessage(event, actors);
            case TYPING_CHANGED -> publishTyping(event, actors);
            case IDLE -> publishAll(actors, actor -> actor.context().boundSession()
                .send(new Messages.ConversationIdleNotify(
                    event.state().conversationId(),
                    event.state()))
                .await());
            case CLOSED -> publishAll(actors, actor -> actor.context().boundSession()
                .send(new Messages.ConversationClosedNotify(
                    event.state().conversationId(),
                    event.state()))
                .await());
        }
    }

    private void publishParticipantJoined(
        ConversationEvent event,
        Map<String, SupportUserActor> actors) {
        String customerActorId = event.state().customerActorId();
        SupportUserActor customer = actors.get(customerActorId);
        if (customer != null && !customer.actorId().equals(event.actorId())) {
            customer.context().boundSession()
                .send(new Messages.ParticipantJoinedNotify(
                    event.state().conversationId(),
                    event.actorId(),
                    event.role(),
                    event.state()))
                .await();
        }
    }

    private void publishAssigned(
        ConversationEvent event,
        Map<String, SupportUserActor> actors) {
        SupportUserActor agent = actors.get(event.actorId());
        if (agent == null) {
            return;
        }
        agent.context().boundSession()
            .send(new Messages.ConversationAssignedNotify(
                event.state().conversationId(),
                event.state()))
            .await();
    }

    private void publishMessage(
        ConversationEvent event,
        Map<String, SupportUserActor> actors) {
        Messages.ChatMessage message = event.message();
        for (SupportUserActor actor : actors.values()) {
            if (actor.actorId().equals(message.senderActorId())) {
                continue;
            }
            actor.context().boundSession()
                .send(new Messages.ChatMessageNotify(
                    event.state().conversationId(),
                    message,
                    event.state()))
                .await();
        }
    }

    private void publishTyping(
        ConversationEvent event,
        Map<String, SupportUserActor> actors) {
        for (SupportUserActor actor : actors.values()) {
            if (actor.actorId().equals(event.actorId())) {
                continue;
            }
            actor.context().boundSession()
                .send(new Messages.TypingChangedNotify(
                    event.state().conversationId(),
                    event.actorId(),
                    Boolean.TRUE.equals(event.isTyping()),
                    event.state()))
                .await();
        }
    }

    private void publishAll(
        Map<String, SupportUserActor> actors,
        Consumer<SupportUserActor> publish) {
        for (SupportUserActor actor : actors.values()) {
            publish.accept(actor);
        }
    }
}

File: framework/languages/java/samples/kotlin/SupportChat/Server/Support/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/support/infrastructure/zlink/notifications/ConversationNotificationPublisher.kt
package systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.notifications

import kotlinx.coroutines.future.await
import systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.actors.SupportUserActor
import systems.zlink.samples.kotlin.supportchat.server.support.domain.conversation.ConversationEvent
import systems.zlink.samples.kotlin.supportchat.server.support.domain.conversation.ConversationEventKind
import systems.zlink.samples.kotlin.supportchat.server.support.domain.conversation.Roles
import systems.zlink.samples.kotlin.supportchat.shared.contracts.ChatMessageNotify
import systems.zlink.samples.kotlin.supportchat.shared.contracts.ConversationAssignedNotify
import systems.zlink.samples.kotlin.supportchat.shared.contracts.ConversationClosedNotify
import systems.zlink.samples.kotlin.supportchat.shared.contracts.ConversationIdleNotify
import systems.zlink.samples.kotlin.supportchat.shared.contracts.ConversationState
import systems.zlink.samples.kotlin.supportchat.shared.contracts.ParticipantJoinedNotify
import systems.zlink.samples.kotlin.supportchat.shared.contracts.TypingChangedNotify

class ConversationNotificationPublisher {
    suspend fun publish(
        events: List<ConversationEvent>,
        actors: Map<String, SupportUserActor>,
    ) {
        for (event in events) {
            publish(event, actors)
        }
    }

    suspend fun publishJoinedAgentToCustomer(
        customer: SupportUserActor,
        state: ConversationState,
    ) {
        val agentActorId = state.agentActorId ?: return
        customer.context().boundSession()
            .send(ParticipantJoinedNotify(state.conversationId, agentActorId, Roles.Agent, state))
            .submit()
            .await()
    }

    private suspend fun publish(
        event: ConversationEvent,
        actors: Map<String, SupportUserActor>,
    ) {
        when (event.kind) {
            ConversationEventKind.PARTICIPANT_JOINED -> publishParticipantJoined(event, actors)
            ConversationEventKind.ASSIGNED -> publishAssigned(event, actors)
            ConversationEventKind.MESSAGE_APPENDED -> publishMessage(event, actors)
            ConversationEventKind.TYPING_CHANGED -> publishTyping(event, actors)
            ConversationEventKind.IDLE ->
                for (actor in actors.values) {
                    actor.context().boundSession()
                        .send(ConversationIdleNotify(event.state.conversationId, event.state))
                        .submit()
                        .await()
                }
            ConversationEventKind.CLOSED ->
                for (actor in actors.values) {
                    actor.context().boundSession()
                        .send(ConversationClosedNotify(event.state.conversationId, event.state))
                        .submit()
                        .await()
                }
        }
    }

    private suspend fun publishParticipantJoined(
        event: ConversationEvent,
        actors: Map<String, SupportUserActor>,
    ) {
        val customer = actors[event.state.customerActorId]
        if (customer != null && customer.actorId() != event.actorId) {
            customer.context().boundSession()
                .send(
                    ParticipantJoinedNotify(
                        event.state.conversationId,
                        event.actorId!!,
                        event.role!!,
                        event.state,
                    ),
                )
                .submit()
                .await()
        }
    }

    private suspend fun publishAssigned(
        event: ConversationEvent,
        actors: Map<String, SupportUserActor>,
    ) {
        val agent = actors[event.actorId] ?: return
        agent.context().boundSession()
            .send(ConversationAssignedNotify(event.state.conversationId, event.state))
            .submit()
            .await()
    }

    private suspend fun publishMessage(
        event: ConversationEvent,
        actors: Map<String, SupportUserActor>,
    ) {
        val message = event.message!!
        for (actor in actors.values) {
            if (actor.actorId() == message.senderActorId) {
                continue
            }
            actor.context().boundSession()
                .send(ChatMessageNotify(event.state.conversationId, message, event.state))
                .submit()
                .await()
        }
    }

    private suspend fun publishTyping(
        event: ConversationEvent,
        actors: Map<String, SupportUserActor>,
    ) {
        for (actor in actors.values) {
            if (actor.actorId() == event.actorId) {
                continue
            }
            actor.context().boundSession()
                .send(
                    TypingChangedNotify(
                        event.state.conversationId,
                        event.actorId!!,
                        event.isTyping == true,
                        event.state,
                    ),
                )
                .submit()
                .await()
        }
    }
}
